sycl/unittests/kernel-and-program/Cache.cpp

```diff
@@ -0,0 +1,521 @@
+public:
+  void operator()(cl::sycl::item<1>){};
+};
+
+class TestKernel2 {
+public:
+  void operator()(cl::sycl::item<1>){};
+};
+
+struct TestCtx {
+  detail::pi::PiContext context;
+};
+
+std::unique_ptr<TestCtx> globalCtx;
+
+static pi_result redefinedProgramCreateWithSource(pi_context context,
+                                                  pi_uint32 count,
+                                                  const char **strings,
+                                                  const size_t *lengths,
+                                                  pi_program *ret_program) {
+  return PI_SUCCESS;
+}
+
+static pi_result
+redefinedProgramBuild(pi_program program, pi_uint32 num_devices,
+                      const pi_device *device_list, const char *options,
+                      void (*pfn_notify)(pi_program program, void *user_data),
+                      void *user_data) {
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedProgramCompile(
+    pi_program program, pi_uint32 num_devices, const pi_device *device_list,
+    const char *options, pi_uint32 num_input_headers,
+    const pi_program *input_headers, const char **header_include_names,
+    void (*pfn_notify)(pi_program program, void *user_data), void *user_data) {
+  return PI_SUCCESS;
+}
+
+static pi_result
+redefinedProgramLink(pi_context context, pi_uint32 num_devices,
+                     const pi_device *device_list, const char *options,
+                     pi_uint32 num_input_programs,
+                     const pi_program *input_programs,
+                     void (*pfn_notify)(pi_program program, void *user_data),
+                     void *user_data, pi_program *ret_program) {
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedKernelCreate(pi_program program,
+                                       const char *kernel_name,
+                                       pi_kernel *ret_kernel) {
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedKernelRetain(pi_kernel kernel) { return PI_SUCCESS; }
+
+static pi_result redefinedKernelRelease(pi_kernel kernel) { return PI_SUCCESS; }
+
+static pi_result redefinedKernelGetInfo(pi_kernel kernel,
+                                        pi_kernel_info param_name,
+                                        size_t param_value_size,
+                                        void *param_value,
+                                        size_t *param_value_size_ret) {
+  if (param_name == PI_KERNEL_INFO_CONTEXT) {
+    auto ctx = reinterpret_cast<detail::pi::PiContext *>(param_value);
+    *ctx = globalCtx->context;
+  }
+
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedKernelSetExecInfo(pi_kernel kernel,
+                                            pi_kernel_exec_info value_name,
+                                            size_t param_value_size,
+                                            const void *param_value) {
+  return PI_SUCCESS;
+}
+TEST(KernelAndProgramCache, ProgramSourceNegativeBuild) {
```

Comment:
Quick Q: why all the tests are negative?

Reply:
There are two positive tests.
These tests are for quite simple cases: build kernel/program with type and with no options.